apis/apiextensions/v1/composition_transforms.go

```diff
@@ -65,6 +72,159 @@ type Transform struct {
 	Convert *ConvertTransform `json:"convert,omitempty"`
 }
 
+// Validate this Transform is valid.
+//
+//nolint:gocyclo // This is a long but simple/same-y switch.
+func (t *Transform) Validate() *field.Error {
+	// TODO(phisco): we should validate each transform type is valid too, e.g. TransformTypeString => String defines the required field for it's type
+	switch t.Type {
+	case TransformTypeMath:
+		if t.Math == nil {
+			return field.Required(field.NewPath("math"), "given transform type math requires configuration")
+		}
+	case TransformTypeMap:
+		if t.Map == nil {
+			return field.Required(field.NewPath("map"), "given transform type map requires configuration")
+		}
+		return xperrors.WrapFieldError(t.Map.Validate(), field.NewPath("map"))
+	case TransformTypeMatch:
+		if t.Match == nil {
+			return field.Required(field.NewPath("match"), "given transform type match requires configuration")
+		}
+	case TransformTypeString:
+		if t.String == nil {
+			return field.Required(field.NewPath("string"), "given transform type string requires configuration")
+		}
+	case TransformTypeConvert:
+		if t.Convert == nil {
+			return field.Required(field.NewPath("convert"), "given transform type convert requires configuration")
+		}
+		if err := t.Convert.Validate(); err != nil {
+			return xperrors.WrapFieldError(err, field.NewPath("convert"))
+		}
+	default:
+		return field.Invalid(field.NewPath("type"), t.Type, "unknown transform type")
+	}
+
+	return nil
+}
+
+type conversionPair struct {
+	from   TransformIOType
+	to     TransformIOType
+	format ConvertTransformFormat
+}
+
+// GetConversionFunc returns the conversion function for the given input and output types, or an error if no conversion is
+// supported. Will return a no-op conversion if the input and output types are the same.
+func (t *ConvertTransform) GetConversionFunc(from TransformIOType) (func(any) (any, error), error) {
+	originalFrom := from
+	to := t.ToType
+	if to == TransformIOTypeInt {
+		to = TransformIOTypeInt64
+	}
+	if from == TransformIOTypeInt {
+		from = TransformIOTypeInt64
+	}
+	if to == from {
+		return func(input any) (any, error) {
+			return input, nil
+		}, nil
+	}
+	f, ok := conversions[conversionPair{from: from, to: to, format: t.GetFormat()}]
+	if !ok {
+		return nil, errors.Errorf("conversion from %s to %s is not supported with format %s", originalFrom, to, t.GetFormat())
+	}
+	return f, nil
+}
+
+// The unparam linter is complaining that these functions always return a nil
+// error, but we need this to be the case given some other functions in the map
+// may return an error.
+var conversions = map[conversionPair]func(any) (any, error){
```

Comment:
We moved it here because we are planning to use it to validate transforms IO types against the schema. We can drop this refactor too, but given that it's a straight forward change I felt like it made sense to have it already in this PR. The motivation behind moving it outside in the original [PR](https://github.com/crossplane/crossplane/pull/3472) was that the reconciler was the only user, which won't be the case anymore as the Composition Validator is also going to use it, so I feel like the precondition do not hold anymore.

Reply:
Apologies for the belated response - I just saw this.

It's not a hill I would die in, but generally I'm not a fan of having a bunch of "business logic" defined as methods on API types. It feels like it mixes a lot of simple type definitions with how they work.

> The motivation behind moving it outside in the original https://github.com/crossplane/crossplane/pull/3472 was that the reconciler was the only user, which won't be the case anymore as the Composition Validator is also going to use it, so I feel like the precondition do not hold anymore.

This was one of the justifications, but I would not say it was the only motivation.